README.md
# MCLD SlideGen
## Automating web slide creation
### Created by Daniel Messer, Web Content Manager

Assembling web slides takes quite a bit of time, and there are often over 100 slides to make every month. Rather than spend hours in Photoshop, I decided to create a web app in PHP, leveraging ImageMagick, that builds the slides programmatically.

## Example Images

**Blue Slide Example**

![Blue slide example](http://i.imgur.com/BUSMIB4.png)

**Grey Slide Example**

![Grey Slide Example](http://i.imgur.com/YvmrrHb.png)

**Purple Slide Example**

![Purple Slide Example](http://i.imgur.com/qygyiao.png)

**Red Slide Example**

![Red Slide Example](http://i.imgur.com/Mvdx6w4.png)

Fonts can be changed within ImageMagick calls and informational bar colours can be altered with Photoshop or GIMP.

## Current workflow

**Step 1. index.php**


First thing accessed by the user. Prompts for upload of the image file. Image file should be at least 780 x 175, but this is not enforced by the
code at this time.

---

**Step 2. imgprep.php**


The user will not see this as it happens in the background. The uploaded image is resized to a width of 780, while keeping the aspect ratio
intact. This not only makes cropping easier, but it also allows more of the image to be cropped in the next step.

---

**Step 3. crop.php**

The newly resized image is presented to the user with a simple cropping interface. The initial size of the cropping box is the exact size of the
image header on the slide. Once the crop is selected...

---

**Step 4. eventdata.php**

This is the data entry screen and is the final screen seen by the user. A simple form allows for entry of the event's title, date and time,
branch, and selection of a bar colour. Once complete, the next step is run, and displayed on the right side of the split screen.

---

**Step 5. createslide.php**

The final step and assembly of the elements. This exectutes behind the scenes and won't be seen by the user. The cropped header from step 3 is
combined with the event data into a final slide. This slide is presented on the right side of the eventdata.php screen.

### Credits

MCLD SlideGen contains code and contributions from:

* Start Bootstrap - [Simple Sidebar.](http://startbootstrap.com/template-overviews/simple-sidebar/) Released under an MIT License.
* Deep Liquid - [Jcrop.](http://deepliquid.com/content/Jcrop.html) Released under an MIT License.
* StackOverflow - [PHP Questions.](http://stackoverflow.com/questions/tagged/php) With gratitude.

Coded in PHP. Utilizing ImageMagick.

* [PHP](http://php.net)
* [ImageMagick](http://www.imagemagick.org)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.4.0] - 2016-04-26

### Added
* Branch selection dropdown
* Automatic file naming based on branch selection dropdown

## [0.3.3] - 2016-03-24

### Added
* Added feature to remove the resized image on the server

## [0.3.2] - 2016-03-18

### Added
* Added a link to allow the user to start creating a new slide after the current one is delivered

## [0.3.1] - 2016-02-25

### Fixed
* Fixed a bug in imgprep.php that caused images to scale incorrectly

## [0.3.0] - 2016-02-23  
### Added
* Jcrop image cropping functionality

### Changed
* Updated image uploading functionality
* Programmatic flow from one part to the next is smoother
* Updated README

## [0.2.1] - 2016-02-23
### Removed
* Removed old files from archive and cleaned up project

## [0.2.1] - 2016-02-19
### Fixed
* Bug that was unaesthetically shrinking dates with double digits

## [0.2.0] - 2016-02-18
### Added
* Bootstrap theme (Simple Sidebar)
* Image uploading
* README file

## [0.1.0] - 2016-02-17  
Initial check in and commits.  

### Added  
* Ability to select the colour of the info bar
* CSS to make the form look nicer
* Initial prep work for file upload


crop.php
<?php
if ($_SERVER['REQUEST_METHOD'] == 'POST')
{
	$targ_w = 740;
	$targ_h = 175;
	$jpeg_quality = 90;

	$src = 'resized.jpg';
	$img_r = imagecreatefromjpeg($src);
	$dst_r = ImageCreateTrueColor( $targ_w, $targ_h );

	imagecopyresampled($dst_r,$img_r,0,0,$_POST['x'],$_POST['y'],
	$targ_w,$targ_h,$_POST['w'],$_POST['h']);

	// header('Content-type: image/jpeg');
	header("Location: eventdata.php");
	imagejpeg($dst_r,null,$jpeg_quality);
	imagejpeg($dst_r, 'headercrop.jpg');
	
		
	
	
	exit;
}

// If not a POST request, display page below:

?>

<html lang="en">
<head>
  <script src="js/jquery.min.js"></script>
  <script src="js/jquery.Jcrop.js"></script>
  <link rel="stylesheet" href="css/main.css" type="text/css" />
  <link rel="stylesheet" href="css/demos.css" type="text/css" />
  <link rel="stylesheet" href="css/jquery.Jcrop.css" type="text/css" />
  
  <script type="text/javascript">

  $(function(){

    $('#cropbox').Jcrop({
      setSelect: [ 0, 0, 740, 175 ],
      onSelect: updateCoords
    });

  });

  function updateCoords(c)
  {
    $('#x').val(c.x);
    $('#y').val(c.y);
    $('#w').val(c.w);
    $('#h').val(c.h);
  };

  function checkCoords()
  {
    if (parseInt($('#w').val())) return true;
    alert('Please select a crop region then press submit.');
    return false;
  };

</script>
<style type="text/css">
  #target {
    background-color: #ccc;
    width: 500px;
    height: 330px;
    font-size: 24px;
    display: block;
  }


</style>

</head>
<body>

<div class="container">
<div class="row">
<div class="span12">
<div class="jc-demo-box">

<div class="page-header">

<h1>Crop image &ndash;</h1>
</div>

		<!-- This is the image we're attaching Jcrop to -->
		<img src="resized.jpg" id="cropbox" />

		<!-- This is the form that our event handler fills -->
		<form action="crop.php" method="post" onsubmit="return checkCoords();">
			<input type="hidden" id="x" name="x" />
			<input type="hidden" id="y" name="y" />
			<input type="hidden" id="w" name="w" />
			<input type="hidden" id="h" name="h" />
			<input type="submit" value="Crop Image" class="btn btn-large btn-inverse" />
		</form>


	</div>
	</div>
	</div>
	</div>
	</body>

</html>


index.php
<!-- This is part of the MCLD SlideGen project -->
<!-- See https://github.com/MCLD/SlideGen for details -->


<!DOCTYPE html>
<html lang="en">

<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>MCLD SlideGen</title>

    <!-- Bootstrap Core CSS -->
    <link href="css/bootstrap.min.css" rel="stylesheet">

    <!-- Custom CSS -->
    <link href="css/simple-sidebar.css" rel="stylesheet">

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
        <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
        <script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
    <![endif]-->

</head>

<body>

    <div id="wrapper">

        <!-- Sidebar -->
        <div id="sidebar-wrapper">

<form action="imgprep.php" method="post" enctype="multipart/form-data">

<h2>SlideGen</h2>
<h3>Upload your image:</h3>

	<span>Select image from computer:</span>
	<input type="file" name="image"><br>

	<button type="submit" >Upload Image</button>
</form>

</div>
</form>

        </div>
        <!-- /#sidebar-wrapper -->

        <!-- Page Content -->
        <div id="page-content-wrapper">
            <div class="container-fluid">
                <div class="row">
                    <div class="col-lg-12">
						<div id="bname"><h1>Your slide will generate here.</h1></div>
                        <iframe name="iframe" width="800" height="400" frameborder="no" border="0"></iframe>
                    </div>
                </div>
            </div>
        </div>
        <!-- /#page-content-wrapper -->

    </div>
    <!-- /#wrapper -->

    <!-- jQuery -->
    <script src="js/jquery.js"></script>

    <!-- Bootstrap Core JavaScript -->
    <script src="js/bootstrap.min.js"></script>

    <!-- Menu Toggle Script -->
    <script>
    $("#menu-toggle").click(function(e) {
        e.preventDefault();
        $("#wrapper").toggleClass("toggled");
    });
    </script>

</div>


</body>

</html>


imgprep.php
<?php

// This is part of the MCLD SlideGen project
// See https://github.com/MCLD/SlideGen for details

// File upload

if(isset($_FILES['image'])){
		$errors= array();
		$file_name = $_FILES['image']['name'];
		$file_size =$_FILES['image']['size'];
		$file_tmp =$_FILES['image']['tmp_name'];
		$file_type=$_FILES['image']['type'];
		$file_ext=strtolower(end(explode('.',$_FILES['image']['name'])));

		$expensions= array("jpeg","jpg","png");
		if(in_array($file_ext,$expensions)=== false){
			$errors[]="extension not allowed, please choose a JPEG or PNG file.";
		}
		if($file_size > 2097152){
		$errors[]='File size must be less than 2 MB';
		}
		if(empty($errors)==true){
			move_uploaded_file($file_tmp,"images/".$file_name);
			echo "Slide created:";
		}else{
			print_r($errors);
		}
	}

exec("convert images/$file_name -resize 780x2000 resized.jpg");

header("Location:crop.php");

exit;
?>


eventdata.php
<!-- This is part of the MCLD SlideGen project -->
<!-- See https://github.com/MCLD/SlideGen for details -->


<!DOCTYPE html>
<html lang="en">

<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>MCLD SlideGen - Event Data</title>

    <!-- Bootstrap Core CSS -->
    <link href="css/bootstrap.min.css" rel="stylesheet">

    <!-- Custom CSS -->
    <link href="css/simple-sidebar.css" rel="stylesheet">

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
        <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
        <script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
    <![endif]-->

</head>

<body>

    <div id="wrapper">

        <!-- Sidebar -->
        <div id="sidebar-wrapper">

<form action="createslide.php" target="iframe" method="post" enctype="multipart/form-data">
<div>

<h2>SlideGen</h2>

<label>
	<span>Event title:</span> <input type="text" name="title"><br><br>
</label>

<label>
	<span>Event date:</span> <input type="text" name="date"><br><br>
</label>

<label>
	<span>Branch:</span>
	<select name="branch">
	<option value="Anderson Branch Library">Anderson</option>
	<option value="Encanto Branch Library">Encanto</option>
	<option value="Fairpoint Branch Library">Fairpoint</option>
	<option value="Freemont Branch Library">Freemont</option>
	<option value="Garfield Branch Library">Garfield</option>
	<option value="Godot Branch Library">Godot</option>
	<option value="Great Lake Branch Library">Great Lake</option>
	<option value="Henderson Branch Library">Henderson</option>
	<option value="Lincoln Branch Library">Lincoln</option>
	<option value="Nirian Branch Library">Nirian</option>
	<option value="North Town Branch Library">North Town</option>
	<option value="Patagonia Branch Library">Patagonia</option>
	<option value="Quinn Branch Library">Quinn</option>
	<option value="Randall Branch Library">Randall</option>
	<option value="South Town Branch Library">South Town</option>
	<option value="Susan Branch Library">Susan</option>
	<option value="Wildlands Branch Library">Wildlands</option>
	</select>
</label>

<label>
	<span>Bar colour:</span>
	<select name="bar">
	<option value="bluebar">Blue</option>
	<option value="greybar">Grey</option>
	<option value="purplebar">Purple</option>
	<option value="redbar">Red</option>
	</select>
</label>

<script>
function setURL(url){
	document.getElementById('iframe').src = url;
}
</script>
<button type="submit" >Generate Slide</button>
<br><br>
<p style="font-size: 10px;"><a href="index.php">Create a new slide.</a></p>
</div>
</form>

        </div>
        <!-- /#sidebar-wrapper -->

        <!-- Page Content -->
        <div id="page-content-wrapper">
            <div class="container-fluid">
                <div class="row">
                    <div class="col-lg-12">
						<div id="bname"><h1>Your slide will generate here.</h1></div>
                        <iframe name="iframe" width="800" height="400" frameborder="no" border="0"></iframe>
                    </div>
                </div>
            </div>
        </div>
        <!-- /#page-content-wrapper -->

    </div>
    <!-- /#wrapper -->

    <!-- jQuery -->
    <script src="js/jquery.js"></script>

    <!-- Bootstrap Core JavaScript -->
    <script src="js/bootstrap.min.js"></script>

    <!-- Menu Toggle Script -->
    <script>
    $("#menu-toggle").click(function(e) {
        e.preventDefault();
        $("#wrapper").toggleClass("toggled");
    });
    </script>

</div>


</body>

</html>


createslide.php
<?php

// This is part of the MCLD SlideGen project
// See https://github.com/MCLD/SlideGen for details


date_default_timezone_set("America/Phoenix");

// Set up variables for the event's title, date, branch, and the colour of the bar.

$title = $_POST['title'];
$date = $_POST['date'];
$branch = $_POST['branch'];
$bar = $_POST['bar'];
$tstamp = date("Ymd-His");

if($branch == 'Anderson Branch Library'):
	$bname = 'an';
	elseif ($branch == 'Encanto Branch Library'):
	$bname = 'en';
	elseif ($branch == 'Fairpoint Branch Library'):
	$bname = 'fa';
	elseif ($branch == 'Freemont Branch Library'):
	$bname = 'fr';
	elseif ($branch == 'Garfield Branch Library'):
	$bname = 'ga';
	elseif ($branch == 'Godot Branch Library'):
	$bname = 'go';
	elseif ($branch == 'Great Lake Branch Library'):
	$bname = 'gr';
	elseif ($branch == 'Henderson Branch Library'):
	$bname = 'he';
	elseif ($branch == 'Lincoln Branch Library'):
	$bname = 'li';
	elseif ($branch == 'Nirian Branch Library'):
	$bname = 'ni';
	elseif ($branch == 'North Town Branch Library'):
	$bname = 'nt';
	elseif ($branch == 'Patagonia Branch Library'):
	$bname = 'pa';
	elseif ($branch == 'Quinn Branch Library'):
	$bname = 'qu';
	elseif ($branch == 'Randall Branch Library'):
	$bname = 'ra';
	elseif ($branch == 'South Town Branch Library'):
	$bname = 'so';
	elseif ($branch == 'Susan Branch Library'):
	$bname = 'su';
	else:
	$bname = 'wt';
endif;

$slidename = "slide-{$bname}-{$tstamp}.png";


// File upload

if(isset($_FILES['image'])){
		$errors= array();
		$file_name = $_FILES['image']['name'];
		$file_size =$_FILES['image']['size'];
		$file_tmp =$_FILES['image']['tmp_name'];
		$file_type=$_FILES['image']['type'];
		$file_ext=strtolower(end(explode('.',$_FILES['image']['name'])));

		$expensions= array("jpeg","jpg","png");
		if(in_array($file_ext,$expensions)=== false){
			$errors[]="extension not allowed, please choose a JPEG or PNG file.";
		}
		if($file_size > 2097152){
		$errors[]='File size must be less than 2 MB';
		}
		if(empty($errors)==true){
			move_uploaded_file($file_tmp,"images/".$file_name);
			echo "Slide created:";
		}else{
			print_r($errors);
		}
	}

// Create slide using ImageMagick

exec("convert -background none -fill white -font \"FreeSerif\" -size 318x70 -gravity West caption:\"".$title."\" eventtitle.png");

exec("convert -background none -fill white -font \"FreeSans\" -size 300x23 -gravity East caption:\"".$date."\" eventdate.png");

exec("convert -background none -fill white -font \"FreeSans\" -size 289x23 -gravity East caption:\"".$branch."\" eventbranch.png");

exec("convert -size 740x250 xc:#a4343a -page +0+0 headercrop.jpg -page +0+175 $bar.png -page +632+175 logo.png -page +15+176 eventtitle.png -page +320+191 eventdate.png -page +330+212 eventbranch.png -layers flatten $slidename");

unlink('resized.jpg');

echo $slidename;

?>

<br>
<img src="<?=$slidename?>">
